Add `See also: `, remove description of other functions in base.array.jl

ChristianKurz opened the pull request:
#23788

Commit:
Add `See also: ` to pop!, shift! and splice!

#23788

KristofferC commented on base/array.jl:
```diff
@@ -1080,6 +1080,8 @@ end
 Remove an item in `collection` and return it. If `collection` is an
 ordered container, the last item is returned.
 
+See also: shift!, splice!
```
```
[`shift!`](@ref)
```

etc

Commit:
added @ref syntax

fredrikekre commented on base/array.jl:
```diff
@@ -1080,6 +1080,8 @@ end
 Remove an item in `collection` and return it. If `collection` is an
 ordered container, the last item is returned.
 
+See also: [`shift!`](@ref), [`splice!`](@ref)
```
What about `push!`

shorty66 replied:
Push! adds elements, whereas the other functions remove elements. I think, `See also:`-lists should be rather minimal: They reduce the need to search the documentation but it's no replacement for a proper search.

StefanKarpinski replied:
A thesaurus lists both similar words and antonyms for a reason – knowing about `push!` is useful if you're looking at `pop!`. From `push!` you could potentially find all the other element addition functions.

garrison commented on base/array.jl:
```diff
@@ -1147,6 +1149,8 @@ end
 
 Remove the first `item` from `collection`.
 
+See also: [`pop!`](@ref), [`splice!`](@ref)
```
What about `unshift!`?

Commit:
Added antonyms, updated docs of append!

`See also`-Lists now contain up to two related functions and antonym at the end

Docs of `append!` were not in line with `prepend!`.

garrison commented on base/array.jl:
````diff
@@ -982,6 +991,10 @@ julia> prepend!([3],[1,2])
 ```
+
+Use [`unshift!`](@ref) to add individual items to `collection` which are not already
+themselves in another collection. The result is of the preceding example is equivalent to
````
I find the phrase "which are not already themselves in another collection" confusing.  I think I know what you are getting at, but I would remove it.

garrison commented on base/array.jl:
```diff
@@ -982,6 +991,10 @@ julia> prepend!([3],[1,2])
+
+Use [`unshift!`](@ref) to add individual items to `collection` which are not already
+themselves in another collection. The result is of the preceding example is equivalent to
+`unshift!([1, 2, 3], 4, 5, 6)`.
```
The previous example (`prepend!([3],[1,2])` I believe) actually has a different result.  And there is a duplicate "is" in the above sentence.

Commit:
Added matching example, corrected sentence

I think the sentence is more clear, with out the "which are not already..." part.

nalimilan commented on base/array.jl:
````diff
@@ -895,9 +898,8 @@ julia> push!([1, 2, 3], 4, 5, 6)
  6
 ```
 
-Use [`append!`](@ref) to add all the elements of another collection to
````
Please don't change unrelated lines. This applies to other places below.

nalimilan commented on base/array.jl:
````diff
@@ -974,14 +978,29 @@ end
+ 3
 ```
+
+Use [`unshift!`](@ref) to add individual items to `collection`. The result of
````
I'd rather put this in the "See also" paragraph, rather than after the examples, where people might not find it (examples can be long). No need to refer to a particular example, just say "`prepend!(x, y)` is equivalent to...".

shorty66 replied:
I removed these sentences in favor of a "See also:" entry. We opted against further descriptions in "See also" earlier.

Commit:
replaced additional descriptions with `See also:`-entry

I removed all descriptions of other functions and inserted an entry in `See also:` instead.
This is related to my first commits' proposal of using a link-description list as `See also:`. 

I also removed the redundant description in `splice!` for removing items. I opted to remove it from the single-item method, as a (negative) range is used to remove items.

fredrikekre commented on base/array.jl:
```diff
@@ -916,9 +916,11 @@ function push!(a::Array{Any,1}, @nospecialize item)
 
 """
-    append!(collection, collection2) -> collection.
+    append!(a::Vector, items) -> collection.
```
Why this change?

shorty66 replied:
`a` has to be a vector for `append!` to work. Of course, `items` could stay `collection`, but this wording is consistent with the other docstrings. Perhaps `collection` would be clearer, though, as items may be interpreted as several seperate items and `append!` expects a collection of items...

```julia
julia> append!([1 2 3], 1)
ERROR: MethodError: no method matching append!(::Array{Int64,2}, ::Int64)

You might have used a 2d row vector where a 1d column vector was required.
Note the difference between 1d column vector [1,2,3] and 2d row vector [1 2 3].
You can convert to a column vector with the vec() function.
Closest candidates are:
  append!(::Array{T,1} where T, ::Any) at array.jl:637

julia> append!([1,2,3], 1)
4-element Array{Int64,1}:
 1
 2
 3
 1
```

nalimilan replied:
This change should go to a separate commit or PR, and it should use `AbstractVector`. I also wonder whether this isn't on purpose, since non-`AbstractVector` collections defined in packages could define methods and it would be silly to repeat the information in another docstring.

fredrikekre commented on base/array.jl:
```diff
@@ -974,13 +972,24 @@ end
  2
  3
+
+julia> prepend!([1, 2, 3], [4, 5, 6])
```
What does this extra example show that the one just above doesn't?

shorty66 replied:
this is a remnant of earlier edits, will remove.

nalimilan commented on base/array.jl:
```diff
@@ -1212,6 +1229,8 @@ end
 Remove the item at the given `i` and return the modified `a`. Subsequent items
 are shifted to fill the resulting gap.
 
+See also: [`splice!`](@ref)
```
Maybe also `pop!` and `unshift!`?

nalimilan commented on base/array.jl:
```diff
@@ -1316,6 +1335,8 @@ Subsequent items are shifted left to fill the resulting gap.
 If specified, replacement values from an ordered
 collection will be spliced in place of the removed item.
 
+See also: [`pop!`](@ref), [`shift!`](@ref), [`insert!`](@ref)
```
Also `deleteat!` (i.e. links should generally go both directions, right?).

shorty66 replied:
Yeah, right. I think this would be easier if `See also:` lists would be generated automatically from some kind of relation-table.

Commit:
removed duplicate examples, added `see also:` items for `deleteat!`
